internal/controller/stages/stages.go

```diff
@@ -706,30 +742,60 @@ func (r *reconciler) syncNormalStage(
-		//       check if verification step is necessary and if yes execute
-		//       step irrespective of phase
-		if (status.Phase == kargoapi.StagePhaseVerifying || status.Phase == kargoapi.StagePhaseNotApplicable) &&
-			stage.Spec.Verification != nil {
-			if status.CurrentFreight.VerificationInfo == nil {
-				if status.Health == nil || status.Health.Status == kargoapi.HealthStateHealthy {
-					log.Debug("starting verification")
-					status.CurrentFreight.VerificationInfo = r.startVerificationFn(ctx, stage)
+		if stage.Spec.Verification != nil {
+			// Confirm if a reverification is requested. If so, clear the
+			// verification info to start the verification process again.
+			info := status.CurrentFreight.VerificationInfo
+			if info != nil && info.ID != "" && info.Phase.IsTerminal() {
+				if v, ok := stage.GetAnnotations()[kargoapi.AnnotationKeyReverify]; ok && v == info.ID {
+					logger.Debug("rerunning verification")
+					status.Phase = kargoapi.StagePhaseVerifying
+					status.CurrentFreight.VerificationInfo = nil
 				}
-			} else {
-				log.Debug("checking verification results")
-				status.CurrentFreight.VerificationInfo = r.getVerificationInfoFn(ctx, stage)
 			}
-			if status.CurrentFreight.VerificationInfo != nil {
-				log.Debugf(
-					"verification phase is %s",
-					status.CurrentFreight.VerificationInfo.Phase,
-				)
-				if status.CurrentFreight.VerificationInfo.Phase.IsTerminal() {
-					// Verification is complete
-					status.Phase = kargoapi.StagePhaseSteady
-					log.Debug("verification is complete")
+
+			// NOTE: If stage cache is stale, phase can be StagePhaseNotApplicable
+			//       even though current freight is not empty in that case
+			//       check if verification step is necessary and if yes execute
+			//       step irrespective of phase
+			if status.Phase == kargoapi.StagePhaseVerifying || status.Phase == kargoapi.StagePhaseNotApplicable {
+				if status.CurrentFreight.VerificationInfo == nil {
+					if status.Health == nil || status.Health.Status == kargoapi.HealthStateHealthy {
+						log.Debug("starting verification")
+						status.CurrentFreight.VerificationInfo = r.startVerificationFn(ctx, stage)
+					}
+				} else {
+					log.Debug("checking verification results")
+					status.CurrentFreight.VerificationInfo = r.getVerificationInfoFn(ctx, stage)
+
+					// Abort the verification if it's still running and the Stage has
+					// been marked to do so.
+					newInfo := status.CurrentFreight.VerificationInfo
+					if newInfo.ID != "" && !newInfo.Phase.IsTerminal() {
+						if v, ok := stage.GetAnnotations()[kargoapi.AnnotationKeyAbort]; ok && v == newInfo.ID {
+							log.Debug("aborting verification")
+							if err := r.abortVerificationFn(ctx, stage); err != nil {
+								return status, fmt.Errorf(
+									"error aborting verification for Stage %q in namespace %q: %w",
+									stage.Name,
+									stage.Namespace,
+									err,
+								)
+							}
```

Comment:
I am on the fence here about returning an error, or returning a `VerificationInfo` struct with the error.

Reply:
> Another interesting thing I have noticed is that if you provide Argo Rollouts with a terminate request, it appears that in some cases the AnalysisRun will end up as Successful. I wonder if this is actually what we want to happen, or if we should construct a custom VerificationInfo with data which reflects the abort operation having taken place.

I think we have separate field for VerificationInfo phase and then, under AnalysRunReference, it has its own phase.

So we should be able to capture that the AnalysisRun succeeded and _still_ record that the verification was aborted.

In determining the VerificationInfo's phase, I agree that the attempt to abort takes precedence over the fact that the attempt raced the AnalysisRun and lost with the AnalysisRun succeeding.

Reply:
> As in the current implementation, we do not distinguish an "object not found" error from e.g. "we temporary can't reach the Kubernetes API server". Which means that even if we could eventually recover from the error, the user has to rerun the verification because the controller has given up on the previous AnalysisRun.

Right. We make no attempt currently to distinguish between errors where a retry helps and one where it doesn't.

I support however you might wish to improve that.

Reply:
> In determining the VerificationInfo's phase, I agree that the attempt to abort takes precedence over the fact that the attempt raced the AnalysisRun and lost with the AnalysisRun succeeding.

It is actually not a race, it is the AnalysisRun reflecting a success while it was the result from a termination request. I am not sure if this is default behavior for Kargo Rollouts, or if it depends on the command which is used to perform the analysis and the potential exit code it gives when it e.g. receives a `SIGKILL`. FWIW, I am testing things with a simple `sleep 600; exit 0;`.

I covered a potential race by checking the current state of the AnalysisRun _before_ adhering to the abort request, which prevents an unnecessary patch operation.

> We make no attempt currently to distinguish between errors where a retry helps and one where it doesn't.
>
> I support however you might wish to improve that.

I think improving this is too big of a change for the current scope of the PR, as it would require bubbling up Kubernetes errors (or translating them into our own typed errors). Where as at present, we have a pattern of translating not found to a `nil` object while returning other errors.

Given this, I am inclined to keep the pattern of reflecting errors in `VerificationInfo` for now. To address this in full in a follow up PR.

Reply:
> It is actually not a race, it is the AnalysisRun reflecting a success while it was the result from a termination request.

Ah. I was mistakenly thinking the AnalysisRun succeeded between the time the abort annotation was added and the time the Stage was reconciled.

This behavior seems peculiar indeed.

@jessesuen would probably be able to give us some insight on this...

@jessesuen you know anything about aborted AnalysisRuns having phase Success?

Reply:
Thinking about this more, this shouldn't block the PR.

As long as the verification phase correctly identifies that the verification was aborted, I don't think the phase of the analysis run, _even if it is arguably incorrect_, matters anywhere else in Kargo.

Reply:
> it appears that in some cases the AnalysisRun will end up as Successful

Is it some cases or all?

> it is the AnalysisRun reflecting a success while it was the result from a termination request

I seem to recall that this was designed behavior for AnalysisRun. There is such thing as an "indefinite" analysisrun that needs to stopped manually / programmatically. 

Let me check on this.

Reply:
Yes, when we terminate an AnalysisRun, we consider it a Successful run:

https://github.com/argoproj/argo-rollouts/blob/master/analysis/analysis.go#L581-L585
https://github.com/argoproj/argo-rollouts/blob/master/analysis/analysis.go#L610-L614

It's because Argo Rollouts will terminate indefinite/background AnalyisRuns when the update is completed. So in the case of Rollouts, termination is a happy path (for background analysis)

I'm thinking about how Kargo should handle this. I think in the case of Kargo, termination is more of a user decision, whereas in Argo Rollouts, the controller deciding to stop the run because the update completed.

Given that, perhaps an terminated analysis run should be treated as failed analysis.